File: aipriceaction/src/models/indicators.rs
/// Calculate Simple Moving Average for a given period.
/// When fewer bars than `period` are available, uses all available data
/// as the window (expanding SMA) so that new tickers still get a value.
pub fn calculate_sma(closes: &[f64], period: usize) -> Vec<f64> {
    let mut ma_values = vec![0.0; closes.len()];

    if period == 0 || closes.is_empty() {
        return ma_values;
    }

    for i in (period - 1)..closes.len() {
        let start_idx = i + 1 - period;
        let sum: f64 = closes[start_idx..=i].iter().sum();
        ma_values[i] = sum / period as f64;
    }

    // Expanding window for the first `period - 1` bars (if dataset is shorter than period)
    let effective_start = if closes.len() < period { 0 } else { period - 1 };
    for i in effective_start..(period - 1).min(closes.len()) {
        let window_len = i + 1;
        let sum: f64 = closes[..=i].iter().sum();
        ma_values[i] = sum / window_len as f64;
    }

    ma_values
}

/// Calculate Weighted Moving Average for a given period.
/// Linear weights: oldest = 1, newest = period.
/// Positions before `period-1` are 0.0.
pub fn calculate_wma(closes: &[f64], period: usize) -> Vec<f64> {
    let mut ma_values = vec![0.0; closes.len()];

    if period == 0 || closes.len() < period {
        return ma_values;
    }

    let weight_sum = period * (period + 1) / 2;

    for i in (period - 1)..closes.len() {
        let start_idx = i + 1 - period;
        let sum: f64 = closes[start_idx..=i]
            .iter()
            .enumerate()
            .map(|(j, &v)| v * (j as f64 + 1.0))
            .sum();
        ma_values[i] = sum / weight_sum as f64;
    }

    ma_values
}

/// Calculate Exponential Moving Average for a given period.
/// Seed: SMA of first `period` closes (or all available if fewer).
/// Then: EMA[i] = close[i] * K + EMA[i-1] * (1-K),  K = 2/(period+1)
pub fn calculate_ema(closes: &[f64], period: usize) -> Vec<f64> {
    let mut ma_values = vec![0.0; closes.len()];
    if period == 0 || closes.is_empty() {
        return ma_values;
    }
    // Seed with SMA of available data (expanding if fewer than period)
    let seed_len = period.min(closes.len());
    let sum: f64 = closes[..seed_len].iter().sum();
    ma_values[seed_len - 1] = sum / seed_len as f64;
    let k = 2.0 / (period as f64 + 1.0);
    for i in seed_len..closes.len() {
        ma_values[i] = closes[i] * k + ma_values[i - 1] * (1.0 - k);
    }
    ma_values
}

/// Calculate MA score: ((close - ma) / ma) * 100
pub fn calculate_ma_score(close: f64, ma: f64) -> f64 {
    if ma == 0.0 {
        0.0
    } else {
        ((close - ma) / ma) * 100.0
    }
}

File: sdk/aipriceaction-python/src/aipriceaction/indicators.py
"""MA indicator calculation, ported from the Rust backend.

Matches src/models/indicators.rs and src/services/aggregator.rs exactly.
"""

from __future__ import annotations

_MA_PERIODS = [10, 20, 50, 100, 200]


def calculate_sma(closes: list[float], period: int) -> list[float]:
    """Simple Moving Average with expanding window for short datasets.

    Mirrors Rust's calculate_sma in src/models/indicators.rs.
    """
    n = len(closes)
    values = [0.0] * n
    if period == 0 or n == 0:
        return values

    # Full window SMA
    for i in range(period - 1, n):
        window = closes[i + 1 - period : i + 1]
        values[i] = sum(window) / period

    # Expanding window for the first `period - 1` bars
    effective_start = 0 if n < period else period - 1
    for i in range(effective_start, min(period - 1, n)):
        values[i] = sum(closes[: i + 1]) / (i + 1)

    return values


def calculate_ema(closes: list[float], period: int) -> list[float]:
    """Exponential Moving Average seeded with SMA.

    Mirrors Rust's calculate_ema in src/models/indicators.rs.
    """
    n = len(closes)
    values = [0.0] * n
    if period == 0 or n == 0:
        return values

    # Seed with SMA of available data
    seed_len = min(period, n)
    values[seed_len - 1] = sum(closes[:seed_len]) / seed_len

    k = 2.0 / (period + 1)
    for i in range(seed_len, n):
        values[i] = closes[i] * k + values[i - 1] * (1.0 - k)

    return values


def calculate_ma_score(close: float, ma: float) -> float:
    """((close - ma) / ma) * 100"""
    if ma == 0.0:
        return 0.0
    return ((close - ma) / ma) * 100.0


def compute_indicators(
    closes: list[float],
    volumes: list[int],
    *,
    use_ema: bool = False,
) -> dict[str, list[float | None]]:
    """Compute all MA indicators and change metrics for a single ticker.

    Mirrors the aggregation logic in Rust's src/services/aggregator.rs.

    Returns a dict with keys: ma10..ma200, ma10_score..ma200_score,
    close_changed, volume_changed, total_money_changed.
    All lists are the same length as closes.
    """
    n = len(closes)
    calc = calculate_ema if use_ema else calculate_sma

    result: dict[str, list[float | None]] = {}

    for period in _MA_PERIODS:
        ma_values = calc(closes, period)
        result[f"ma{period}"] = [v if v > 0.0 else None for v in ma_values]
        result[f"ma{period}_score"] = [
            calculate_ma_score(closes[i], v) if v > 0.0 else None
            for i, v in enumerate(ma_values)
        ]

    # Change metrics (need previous bar)
    close_changed: list[float | None] = [None] * n
    volume_changed: list[float | None] = [None] * n
    total_money_changed: list[float | None] = [None] * n

    for i in range(1, n):
        prev_close = closes[i - 1]
        prev_vol = volumes[i - 1]

        if prev_close > 0.0:
            close_changed[i] = ((closes[i] - prev_close) / prev_close) * 100.0

        if prev_vol > 0:
            volume_changed[i] = ((volumes[i] - prev_vol) / prev_vol) * 100.0

        total_money_changed[i] = (closes[i] - prev_close) * volumes[i]

    result["close_changed"] = close_changed
    result["volume_changed"] = volume_changed
    result["total_money_changed"] = total_money_changed

    return result
